File: GameObjects.py
#Define physics object

import math
import pyglet
from pyglet.window import key
from GameResources import *

class PhysicsObject(pyglet.sprite.Sprite):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.vx = 0.0
        self.vy = 0.0

        self.radius = 0.0

        self.colliding_with = []
        self.collides_with = []

        self.dead = False

        self.new_objects = []

    def calculate_distance(self,obj):
        return math.sqrt((self.x-obj.x)**2 + (self.y-obj.y)**2)

    def check_bounds(self):
        min_x = self.radius
        min_y = self.radius
        max_x = 800-self.radius
        max_y = 600-self.radius

        if self.x < min_x:
            self.x = min_x
        if self.y < min_y:
            self.y = min_y
        if self.x > max_x:
            self.x = max_x
        if self.y > max_y:
            self.y = max_y

    def check_collisions(self, objects):
        for obj in objects:
            if self.calculate_distance(obj) <= self.radius+obj.radius:
                self.colliding_with.append(obj)
                obj.colliding_with.append(self)

    def check_death(self):
        for obj in self.colliding_with:
            if type(obj) in self.collides_with:
                self.dead = True

    def update(self,dt,objects):
        self.x += self.vx*dt
        self.y += self.vy*dt

        self.check_bounds()
        self.check_collisions(objects)
        self.check_death()
        self.colliding_with = []

class Player(PhysicsObject):

    def __init__(self, *args, **kwargs):
        super().__init__(img=player_image, *args, **kwargs)

        self.scale = 1.0
        self.radius = 10

        self.key_handler = key.KeyStateHandler()

        self.collides_with = [Enemy]

        self.energy = 100
        self.bullet_cost = 10
        self.energy_bar = pyglet.text.Label(text="Energy: " + str(self.energy),x=10,y=10,batch=self.batch)
    
    def fire_bullet(self):
        if self.energy >= self.bullet_cost:
            self.energy -= self.bullet_cost

            bullet = Bullet(x=self.x,y=self.y,batch=self.batch)
            bullet.vy = self.vy
            bullet.vx = 500 + self.vx
            self.new_objects.append(bullet) 

    def update(self, dt, objects):


        self.energy += 0.3333333333
        if self.energy > 100:
            self.energy = 100

        self.energy_bar.text = "Energy: " + str(int(self.energy))

        #self.vx = 0
        #self.vy = 0

        self.vx *= 0.9
        self.vy *= 0.9

        if self.key_handler[key.W]:
            self.vy += 20
        if self.key_handler[key.S]:
            self.vy -= 20
        if self.key_handler[key.D]:
            self.vx += 20
        if self.key_handler[key.A]:
            self.vx -= 20


        v = math.sqrt(self.vx**2 + self.vy**2)
        if v > 300.0:
            self.vx = 300*self.vx/v
            self.vy = 300*self.vy/v

        #self.vx *= 200
        #self.vy *= 200

        super(Player, self).update(dt, objects)

    def on_key_press(self, symbol, modifiers):
        if symbol == key.SPACE:
            self.fire_bullet()

class Enemy(PhysicsObject):
    
    def __init__(self, *args, **kwargs):
        super().__init__(img=enemy_image, *args, **kwargs)

        self.scale = 0.2
        self.radius = 25

        self.collides_with = [Bullet]


    def check_bounds(self):
        min_x = self.radius
        min_y = self.radius
        max_x = 800-self.radius
        max_y = 600-self.radius

        if self.x < min_x:
            self.dead = True
        if self.y < min_y:
            self.y = min_y
        if self.x > max_x:
            self.x = max_x
        if self.y > max_y:
            self.y = max_y

class Bullet(PhysicsObject):

    def __init__(self, *args, **kwargs):
        super().__init__(img = bullet_image, *args, **kwargs)

        self.scale = 1.0

        self.radius = 8

        self.collides_with = [Enemy]

    def check_bounds(self):
        min_x = self.radius
        min_y = self.radius
        max_x = 800-self.radius
        max_y = 600-self.radius

        if self.x < min_x:
            self.dead = True
        if self.y < min_y:
            self.dead = True
        if self.x > max_x:
            self.dead = True
        if self.y > max_y:
            self.dead = True

File: GameObjects.js
import * as THREE from 'three';
import { KeyboardState } from './KeyHandler';
import { loadObject } from './LoadingHandler';

export class PhysicsObject extends THREE.Sprite {

    constructor(material) {
        super(material)  ;

        this.vx = 0.0;
        this.vy = 0.0;

        this.radius = 0.0;

        this.colliding_with = [];
        this.collides_with = [];

        this.dead = false;
        this.death_sound = null;

        this.new_objects = [];


        this.died_to_bullet = false;

    }

    calculate_distance(obj){
        return Math.sqrt((this.position.x-obj.position.x)**2 + (this.position.y-obj.position.y)**2);

    }

    check_bounds(){
        var min_x = this.radius-window.innerWidth/2;
        var min_y = this.radius-window.innerHeight/2;
        var max_x = window.innerWidth/2-this.radius;
        var max_y = window.innerHeight/2-this.radius;

        if (this.position.x < min_x){
            this.position.x = min_x;
            this.vx = 0;
        }
        if (this.position.y < min_y){
            this.position.y = min_y;
            this.vy = 0;
        }
        if (this.position.x > max_x){
            this.position.x = max_x;
            this.vx = 0;
        }
        if (this.position.y > max_y){
            this.position.y = max_y;
            this.vy = 0;
        }
    }

    check_collisions(objects){
        objects.forEach(obj => {
            if (this.calculate_distance(obj) <= this.radius+obj.radius){
                this.colliding_with.push(obj);
                obj.colliding_with.push(this);
            }            
        });
    }

    check_death(){
        this.colliding_with.forEach((obj) => {
            this.collides_with.forEach((collision_target) => {
                if (obj instanceof collision_target) {
                    this.dead = true;
                    if (obj instanceof Bullet){
                        this.died_to_bullet = true;
                    }
                    if (this.death_sound !== null){
                        this.death_sound.play();
                    }
                }
            });            
        });
    }

    update(dt,objects){
        this.position.x += this.vx*dt;
        this.position.y += this.vy*dt;

        this.check_bounds();
        this.check_collisions(objects);
        this.check_death();
        this.colliding_with = [];
    }

}

export class Player extends PhysicsObject{

    constructor(player_image){
        super(player_image);

        this.position.z = 0.01;

        this.radius = 16;
        this.max_speed = 0.25;
        this.acceleration = 0.005;

        this.key_handler = new KeyboardState();
        this.key_handler.push_handler(this);

        this.collides_with = [Enemy];

        this.energy = 100;
        this.energy_label = document.getElementById("energy");
        this.energy_label.textContent = "Energy: "+Math.ceil(this.energy);


        this.bullet_cost = 10;
        this.bullet_speed = 0.65;

        this.new_objects = [];

        this.bullet_sound = new Howl({
            src: ['sfx/584196__unfa__weapons-plasma-shot-04.flac']
        });

        this.death_sound = new Howl({
            src: ['sfx/435413__v-ktor__explosion12.wav']
        })
    }


    async fire_bullet(){
        if (this.energy > this.bullet_cost){
            this.energy -= this.bullet_cost;
            

            const bullet = await loadObject('sprites/Nasa_Bullet_Sprite.png', Bullet);
            bullet.vy = this.vy;
            bullet.vx = this.bullet_speed + this.vx;
            bullet.position.x = this.position.x;
            bullet.position.y = this.position.y;
            this.new_objects.push(bullet);

            this.bullet_sound.play();
        }
    }

    update( dt, objects){


        this.energy += 0.3333333333;
        if (this.energy > 100){
            this.energy = 100;
        }

        this.energy_label.textContent = "Energy: "+Math.ceil(this.energy);

        this.vx *= 0.9;
        this.vy *= 0.9;

        if (this.key_handler.isKeyPressed('KeyW')){
            this.vy += dt*this.acceleration;
        }
        if (this.key_handler.isKeyPressed('KeyS')){
            this.vy -= dt*this.acceleration;
        }
        if (this.key_handler.isKeyPressed('KeyD')){
            this.vx += dt*this.acceleration;
        }
        if (this.key_handler.isKeyPressed('KeyA')){
            this.vx -= dt*this.acceleration;
        }


        var v = Math.sqrt(this.vx**2 + this.vy**2);
        if (v > this.max_speed){
            this.vx = this.max_speed*this.vx/v;
            this.vy = this.max_speed*this.vy/v;
        }

        super.update(dt, objects);
    }

    onKeyDown(event){
        if (event.code === 'Space'){
            if (this.dead === false) {
                this.fire_bullet();
            }
        }
    }
    onKeyUp(event){
        return
    }
}

export class Bullet extends PhysicsObject{

    constructor(bullet_image){
        super(bullet_image);

        this.radius = 8;

        this.collides_with = [Enemy];
    }

    check_bounds(){
        var min_x = this.radius-window.innerWidth/2;
        var min_y = this.radius-window.innerHeight/2;
        var max_x = window.innerWidth/2-this.radius;
        var max_y = window.innerHeight/2-this.radius;

        if (this.position.x < min_x){
            this.dead = true;
        }
        if (this.position.y < min_y){
            this.dead = true;
        }
        if (this.position.x > max_x){
            this.dead = true;
        }
        if (this.position.y > max_y){
            this.dead = true;
        }
    }
}

export class Enemy extends PhysicsObject{
    constructor(enemy_image){
        super(enemy_image);

        this.radius = 20;

        this.collides_with = [Bullet];

        this.death_sound = new Howl({
            src: ['sfx/435413__v-ktor__explosion12.wav']
        });
    };
    


    check_bounds(){
        var min_x = this.radius-window.innerWidth/2;
        var min_y = this.radius-window.innerHeight/2;
        var max_x = window.innerWidth/2-this.radius;
        var max_y = window.innerHeight/2-this.radius;

        if (this.position.x < min_x){
            this.dead = true;
        }
        if (this.position.y < min_y){
            this.position.y = min_y;
        }
        if (this.position.x > max_x){
            this.position.x = max_x;
        }
        if (this.position.y > max_y){
            this.position.y = max_y;
        }
    }
}

export class TrackingEnemy extends Enemy{
    constructor(enemy_image){
        super(enemy_image);

        this.radius = 20;

        this.collides_with = [Bullet];

        this.death_sound = null;

        this.target = null;
        this.tracking_strength = 1.0

        this.walk_speed = 0.1
        this.charge_speed = 0.2
    }

    track_player(target) {
        var target_direction_x = target.position.x - this.position.x
        var target_direction_y = target.position.y - this.position.y
        var target_distance = Math.sqrt(target_direction_x**2+target_direction_y**2)
        target_direction_x /= target_distance
        target_direction_y /= target_distance

        return [target_direction_x, target_direction_y, target_distance]
    }


    check_bounds(){
        var min_x = this.radius-window.innerWidth/2;


        if (this.position.x < min_x){
            this.dead = true;
        }
        
        super.check_bounds();
    }

    update(dt,objects) {
        this.vx -= 0.01

        if (this.target !== null) {
            var [player_direction_x, player_direction_y, player_distance] = this.track_player(this.target)
            if (player_distance < 50) {
                this.vx = this.charge_speed*player_direction_x
                this.vy = this.charge_speed*player_direction_y
                this.target=null
            } else {
                this.vx += this.tracking_strength*player_direction_x
                this.vy += this.tracking_strength*player_direction_y
            }
        }
        
        var v = Math.sqrt(this.vx**2 + this.vy**2)
        if (this.target !== null) {
            if (v > this.walk_speed) {
                this.vx = this.walk_speed*this.vx/v
                this.vy = this.walk_speed*this.vy/v
            }
        } else {
            if (v > this.charge_speed) {
                this.vx = this.charge_speed*this.vx/v
                this.vy = this.charge_speed*this.vy/v
            }
        }

        super.update(dt,objects)
    }
}





